cloud/log_syncer.go

```diff
@@ -158,27 +151,58 @@ func (s *LogSyncer) enqueue(l *DeploymentLog) {
 	s.lastEnqueued = time.Now()
 }
 
-// scanLines is a split function for a [bufio.Scanner] that returns each line of
-// text. It's similar to [bufio.ScanLines] but do not remove the trailing newline
-// marker and optional carriege return. The returned line may be empty.
-// The end-of-line marker is one optional carriage return followed
-// by one mandatory newline. In regular expression notation, it is `\r?\n`.
-// The last non-empty line of input will be returned even if it has no
-// newline.
-func scanLines(data []byte, atEOF bool) (advance int, token []byte, err error) {
-	if atEOF && len(data) == 0 {
-		return 0, nil, nil
-	}
-	if i := bytes.IndexByte(data, '\n'); i >= 0 {
-		// We have a full newline-terminated line.
-		return i + 1, data[0 : i+1], nil
+func readLines(r io.Reader, pending []byte) (line [][]byte, rest []byte, err error) {
+	const readSize = 1024
+
+	var buf [readSize]byte
+	rest = pending
+	for {
+		n, err := r.Read(buf[:])
+		if n > 0 {
+			rest = append(rest, buf[:n]...)
+			var lines [][]byte
+
+			var nlpos int
+			for nlpos != -1 {
+				nlpos = bytes.IndexByte(rest, '\n')
+				if nlpos >= 0 {
+					lines = append(lines, rest[:nlpos+1]) // line includes ln
+					rest = rest[nlpos+1:]
+				}
+			}
+			if len(lines) > 0 {
+				return lines, rest, err
+			}
+		} else if err == nil {
+			// misbehaving reader
+			return nil, nil, io.EOF
+		}
+		if err != nil {
+			return nil, rest, err
+		}
+
+		// line ending not found, continue reading.
 	}
-	// If we're at EOF, we have a final, non-terminated line. Return it.
-	if atEOF {
-		return len(data), data, nil
+}
+
+func readLine(r io.Reader) (line []byte, err error) {
```

Comment:
I think this could be removed later, the readLines implementation looks universally better to me and it's not much more complicated, so I don't think we will often need this benchmark in the future.